### PR TITLE
Fix Ephemeral-emptyDir when set volumeEnabled: false

### PR DESCRIPTION
**Pull request by robinson**
Fix the error validating data: [ValidationError(Pod.spec.volumes[0].ephemeral): unknown field "emptyDir" in io.k8s.api.core.v1.EphemeralVolumeSource

### CODE REVIEW

**Comment by clemlesne** [comment]
Thank you for for that great fix.

I initially had the idea to use ephemeral storage for all scenarios. Some Kubernetes setups are not compatible with ephemeral disks, and I thought ephemeral emptyDir was a nice idea as a workaround, even with size / performance limitations, and evictions risks.

Functionally, it is required to clean the disk after each pipeline execution. The security breach is enormous if data or context is shared between pipelines.

emptyDir are reused across Pod restart, but not across Pods. But by security it is already explicitly defined to never restart a Pod.

I propose to add an explicit warning in addition to the restart deny, to alert a future user of this necessity.

File: src/helm/azure-pipelines-agent/templates/_helpers.tpl [hunk 1]
```diff
@@ -188,13 +188,13 @@ volumes:
...
               storage: {{ .Values.pipelines.cache.size | required "A value for .Values.pipelines.cache.size is required" }}
       {{- else }}
-      emptyDir:
-        sizeLimit: {{ .Values.pipelines.cache.size | required "A value for .Values.pipelines.cache.size is required" }}
+    emptyDir:
+      sizeLimit: {{ .Values.pipelines.cache.size | required "A value for .Values.pipelines.cache.size is required" }}
       {{- end }}
   {{- if not .Values.image.isWindows }}
   - name: local-tmp
+  {{- if .Values.pipelines.tmpdir.volumeEnabled }}
```
Comment (clemlesne) [inline src/helm/azure-pipelines-agent/templates/_helpers.tpl:196]: Missing one indentation

File: src/helm/azure-pipelines-agent/templates/_helpers.tpl [hunk 2]
```diff
@@ -178,8 +178,8 @@ containers:
       {{- end }}
 volumes:
   - name: azp-work
+        {{- if .Values.pipelines.cache.volumeEnabled }}
```
Comment (clemlesne) [inline src/helm/azure-pipelines-agent/templates/_helpers.tpl:181]: Missing one indentation

**Review by clemlesne — CHANGES_REQUESTED** [review]

**Comment by clemlesne** [comment]
> emptyDir are reused across Pod restart, but not across Pods. But by security it is already explicitly defined to never restart a Pod.
> 
> I propose to add an explicit warning in addition to the restart deny, to alert a future user of this necessity.

All, [excepted the installation without KEDA installed, based on Deployment](https://github.com/clemlesne/azure-pipelines-agent/blob/a8e46abcfe6d5d853527030f011097e165f80c1a/src/helm/azure-pipelines-agent/templates/_helpers.tpl#LL101C1-L101C1)… **So, we need a solution to wipe the disk.**

First, could you test if the command `config.sh remove` already wipes the disk? The doc is unclear about it.

If not, what I propose you is to add an ENV to the Pod, only for `restartPolicy` != `Never`, enabling a disk cleanup at each restart (solution needs to be decided).

**Comment by clemlesne** [comment]
The approach I can propose you is based on https://github.com/Microsoft/azure-pipelines-agent/issues/1895#issuecomment-430610732.

Using the `preStop` hook to clean the directory, like so:

```bash
bash config.sh remove --auth PAT --token ${AZP_TOKEN}; rm -rf ${AZP_WORK}
```

Complete example in Helm:

```yaml
lifecycle:
      preStop:
        exec:
          {{- if .Values.image.isWindows }}
          command: [powershell, -Command, ".\\config.cmd remove --auth PAT --token $Env:AZP_TOKEN; Remove-Item -Recurse -Force $Env:AZP_WORK"]
          {{- else }}
          command: [bash, -c, "bash config.sh remove --auth PAT --token ${AZP_TOKEN}; rm -rf ${AZP_WORK}"]
          {{- end }}
```

**Comment by robinson** [comment]
Hi @clemlesne sorry for the lately response. The issue is: when set _volumeEnabled = true_, then the volume should be:
 - name: local-tmp
     ephemeral:
      volumeClaimTemplate:
        spec:
          accessModes: [ "ReadWriteOnce" ]
...
When set _volumeEnabled = false_ it should be
 - name: local-tmp
     **ephemeral:
      emptyDir:**
        sizeLimit: 1000
     
 then, the error as I show above. Which _emptyDir_ is not belonging to _ephemeral_???
so, my fix in case set _volumeEnabled = false_ it should be
 - name: local-tmp
      emptyDir:
        sizeLimit: 1000
It worked fine with my k8s rke. I don't know whether it would work for other k8s setup.
refer: https://kubernetes.io/docs/concepts/storage/ephemeral-volumes/
Thanks!
Robinson

**Comment by clemlesne** [comment]
> Functionally, it is required to clean the disk after each pipeline execution. The security breach is enormous if data or context is shared between pipelines.
> 
> emptyDir are reused across Pod restart, but not across Pods.

emptyDir without wipe at Pod restart (what does ephemeral storage) creates a security risk.

> Using the `preStop` hook to clean the directory, like so:
> 
> ```bash
> bash config.sh remove --auth PAT --token ${AZP_TOKEN}; rm -rf ${AZP_WORK}
> ```
> 
> Complete example in Helm:
> 
> ```yaml
> lifecycle:
>   preStop:
>     exec:
>       {{- if .Values.image.isWindows }}
>       command: [powershell, -Command, ".\\config.cmd remove --auth PAT --token $Env:AZP_TOKEN; Remove-Item -Recurse -Force $Env:AZP_WORK"]
>       {{- else }}
>       command: [bash, -c, "bash config.sh remove --auth PAT --token ${AZP_TOKEN}; rm -rf ${AZP_WORK}"]
>       {{- end }}
> ```

This is the mitigation I propose.

What do you think?

**Comment by robinson** [comment]
I have tested, it worked for me. Thanks

**Pull request closed without merging**